src/ProtectedRooms.ts

```diff
@@ -0,0 +1,518 @@
+
+    /**
+     * This is a queue for redactions to process after mjolnir
+     * has finished applying ACL and bans when syncing.
+     */
+    private readonly eventRedactionQueue = new EventRedactionQueue();
+
+    private readonly errorCache = new ErrorCache();
+
+    /**
+     * These are globs that are matched against the reason of an `m.ban` recommendation against a user.
+     * If a rule matches a user in a room, and a glob from here matches that rule's reason, then we will redact
+     * all of the messages from that user.
+     */
+    private automaticRedactionReasons: MatrixGlob[] = [];
+
+    /**
+     * Used to provide mutual exclusion when synchronizing rooms with the state of a policy list.
+     * This is because requests operating with rules from an older version of the list that are slow
+     * could race & give the room an inconsistent state. An example is if we add multiple m.policy.rule.server rules,
+     * which would cause several requests to a room to send a new m.room.server_acl event.
+     * These requests could finish in any order, which has left rooms with an inconsistent server_acl event
+     * until Mjolnir synchronises the room with its policy lists again, which can be in the region of hours.
+     */
+    private aclChain: Promise<void> = Promise.resolve();
+
+    constructor(
+        private readonly client: MatrixClient,
+        private readonly clientUserId: string,
+        private readonly managementRoomId: string,
+        private readonly managementRoomOutput: ManagementRoomOutput,
+        /**
+         * The protection manager is only used to verify the permissions
+         * that the protection manager requires are correct for this set of rooms.
+         * The protection manager is not really compatible with this abstraction yet
+         * because of a direct dependency on the protection manager in Mjolnir commands.
+         */
+        private readonly protectionManager: ProtectionManager,
+        private readonly config: IConfig,
+    ) {
+        for (const reason of this.config.automaticallyRedactForReasons) {
+            this.automaticRedactionReasons.push(new MatrixGlob(reason.toLowerCase()));
+        }
+
+        // Setup room activity watcher
+        this.protectedRoomActivityTracker = new ProtectedRoomActivityTracker();
+    }
+
+    /**
+     * Queue a user's messages in a room for redaction once we have stopped synchronizing bans
+     * over the protected rooms.
+     *
+     * Nit: it probably should just called RedactUser and the queue is a hidden detail but whatever.
+     * @param userId The user whose messages we want to redact.
+     * @param roomId The room we want to redact them in.
+     */
+    public queueRedactUserMessagesIn(userId: string, roomId: string) {
+        this.eventRedactionQueue.add(new RedactUserInRoom(userId, roomId));
+    }
+
+    public get automaticRedactGlobs(): MatrixGlob[] {
```

Comment:
Nit: If this needs to be `public`, could you copy the documentation?

Also, could this be `ReadOnly<MatrixGlob[]>`?